### DISCUSSION TITLE
Modbus over UDP

### DISCUSSION THREAD

**Discussion by canton7** [nathanmarlor/foxess_modbus#75] [General]
Hi @nathanmarlor,

I'm thinking of adding support for UDP, as that seems to work much better with my W610.

The most obvious thing to do is update the config flow to show:

* TCP
* UDP
* Serial

However, I wonder whether that's getting a bit unintuitive. If someone's connected a LAN cable directly into their inverter, they don't necessarily know whether the inverter wants to talk TCP or UDP over that cable, and knowing the difference might be beyond them.

Instead, I was wondering whether it might make sense to do something like:

* LAN cable into inverter
* RS-485 adapter using serial
* RS-485 adapter using TCP
* RS-485 adapter using UDP

Hopefully the user knows how they're connected to the inverter, and if they're using an RS-465 adapter maybe they saw "tcp" or "udp" in the settings somewhere when they were setting it up.

That also means we don't need to do auto-detection of the connection type.

Do you have any thoughts? I'm happy to hack away here, but I wanted to run things past you as this is an area you've put a lot of effort into!

**Comment by nathanmarlor** [comment]
Thanks @canton7 !

Agree with your points - we need to find a balance between 'advanced' options such as TCP/UDP (+ potentially more in the future) and making the config flow intuitive for the less technically minded. As some form of compromise could we add a selector with Autodetect / TCP 485 / UDP 485 / LAN? Autodetect would run the existing routine, whilst the TCP etc. would effectively act as overrides if you have a known setup.

Whatd'ya think?

**Reply by canton7** [reply]
We can't have an "Autoconnect" on its own, as we do need an IP address / serial device. So we need some way to select between "Ethernet" and "Serial". However, you're right in that we can probably auto-detect between TCP and UDP quite happily.

How about the question "How are you connected to your inverter?" with the options:

* Direct LAN connection
* RS-484 adapter using LAN
* RS-485 adapter using serial

Selecting "RS-484 adapter using LAN" gives you a page with:

* Protocol: Comobox with Autodetect / TCP / UDP, default to Autodetect
* IP Address
* Port

The reason for breaking out "Direct LAN connection" vs "RS-484 adapter using LAN", even though we can auto-detect between them, is that we'd need a term for "Anything over TCP or UDP", and saying "Direct LAN connection or RS-485 adapter using LAN" is a bit of a mouthful: if the user knows they've plugged in a LAN cable but doesn't know what RS-485 is, then we've confused things by mashing together something they know with something they're confused by.

**Reply by nathanmarlor** [reply]
Sounds great - even though the autodetect routine becomes redundant I'd be tempted to leave it in as a connection check? It's also useful to grab the inverter model (and possibly serial # in the future)?

**Reply by canton7** [reply]
Although thinking about it more -- maybe the thing to do is to word it something like "How does HomeAssistant connect to your inverter", with the options:

* Ethernet cable
* Serial device

So, phrase it in terms of what's connected into the computer which HA is running on. Then on the "Ethernet cable" page, we can have that combobox for Autodetect/ModbusTCP/ModbusUDP" or similar (with TCP still running auto-detected to figure out whether it's LAN or AUX), just in case they've got a modbus<->RS485 bridge which accepts both TCP and UDP, and they want to force the issue.

**Reply by canton7** [reply]
Although having said that, I'm not sure I'd know which one to pick myself there. As I know that RS-485 is serial, so my W610 talks both with LAN *and* serial... Hmph.

**Reply by canton7** [reply]
> even though the autodetect routine becomes redundant I'd be tempted to leave it in as a connection check? It's also useful to grab the inverter model (and possibly serial # in the future)?

Yep absolutely. We still need to auto-detect the model as you say, and in one of those combinations of options we'd auto-detect between TCP and UDP.

**Reply by nathanmarlor** [reply]
> Although thinking about it more -- maybe the thing to do is to word it something like "How does HomeAssistant connect to your inverter", with the options:
> 
> * Ethernet cable
> * Serial device
> 
> So, phrase it in terms of what's connected into the computer which HA is running on. Then on the "Ethernet cable" page, we can have that combobox for Autodetect/ModbusTCP/ModbusUDP" or similar (with TCP still running auto-detected to figure out whether it's LAN or AUX), just in case they've got a modbus<->RS485 bridge which accepts both TCP and UDP, and they want to force the issue.

I think this is the way to go - if HA weren't so brittle I'd consider including two small graphics next to the options to give a visual representative too. Perhaps it's worth throwing a couple of options into the Fox Facebook group to see where the general consensus lies?

**Reply by canton7** [reply]
Thinking about it more, I think the options "Ethernet cable" and "USB cable" are maybe better? Since those serial-to-RS485 adapters connect using a USB cable.

**Reply by canton7** [reply]
Bouncing back off #40 -- maybe the first option should be a drop-down / set of radio buttons to select:

* Direct LAN connection to inverter
* USR W610
* Waveshare
* Whatever that USB one is that people use
* Other Ethernet to RS-485 adapter
* Other USB to RS-485 adapter 

We can use that to figure out whether to prompt for IP or serial device, and what defaults to apply for things like max fetch window.

That's also nice and recognisable -- I don't need to know that my W610 uses an ethernet cable or anything, I just pick it from the list
